File: game/Assets/Scripts/Network/Messages/Action.cs
using System;
using UnityEngine;

namespace Network
{
    [Serializable]
    public enum ActionType
    {
        INIT = 0,
        INSTANT,
        ROTATE,
        TRANSLATE,
        OUTLINE,
        DEATH,
    }

    [Serializable]
    public enum AnimationType
    {
        NONE = 0,
        IDLE,
        WALKING,
        INSTANT,
        TRANSLATE,
        ACCEL_DECEL,
        SKIPPING,
        ROTATE,
    }

    // Color is 
    [Serializable]
    public class Color
    {
        // Some defaults for convenience.
        public static readonly Color White = new Color(1, 1, 1, 1);
        public static readonly Color Red = new Color(1, 0, 0, 1);
        public static readonly Color Green = new Color(0, 1, 0, 1);
        public static readonly Color Blue = new Color(0, 0, 1, 1);
        public static readonly Color Black = new Color(0, 0, 0, 1);
        public static readonly Color Yellow = new Color(1, 0.92f, 0.016f, 1);
        public static readonly Color Magenta = new Color(1, 0, 1, 1);
        public static readonly Color Cyan = new Color(0, 1, 1, 1);
        public static readonly Color Grey = new Color(0.5f, 0.5f, 0.5f, 1);
        public static readonly Color Transparent = new Color(0, 0, 0, 0);

        public Color(float r_init, float g_init, float b_init, float a_init)
        {
            r = r_init;
            g = g_init;
            b = b_init;
            a = a_init;
        }

        // Of course, a Unity Color conversion function.
        public UnityEngine.Color ToUnity()
        {
            return new UnityEngine.Color(r, g, b, a);
        }

        public override string ToString()
        {
            return string.Format("({0}, {1}, {2}, {3})", r, g, b, a);
        }

        // RGBA. Floating point. 0 to 1.
        public float r;
        public float g;
        public float b;
        public float a;
    }

    [Serializable]
    public class Action
    {
        public int id;
        public ActionType action_type;
        public AnimationType animation_type;
        public HecsCoord displacement;
        public float rotation;  // Degrees.
        public float border_radius;  // Outline radius.
        public Color border_color;  // Outline color.
        public float opacity;  // Used to animate some UI elements.
        public float duration_s;
        public string expiration;  // DateTime in ISO 8601.
        public Color border_color_follower_pov;  // Outline color from follower's POV.
    }
}  // namespace Network

File: server/messages/action.py
import logging
from dataclasses import dataclass, field, replace
from datetime import datetime, timedelta
from enum import Enum
from typing import Optional

from dateutil import tz
from mashumaro import pass_through
from mashumaro.mixins.json import DataClassJSONMixin

from server.hex import HecsCoord

logger = logging.getLogger()

DEFAULT_WALK_TIME_S = 0.45
DEFAULT_TURN_TIME_S = 0.33


class ActionType(Enum):
    INIT = 0
    INSTANT = 1
    ROTATE = 2
    TRANSLATE = 3
    OUTLINE = 4
    DEATH = 5
    NONE = 6


class AnimationType(Enum):
    NONE = 0
    IDLE = 1
    WALKING = 2
    INSTANT = 3
    TRANSLATE = 4
    ACCEL_DECEL = 5
    SKIPPING = 6
    ROTATE = 7


@dataclass(frozen=True)
class Color(DataClassJSONMixin):
    r: float
    g: float
    b: float
    a: float

    def __eq__(self, rhs):
        return (
            self.r == rhs.r and self.g == rhs.g and self.b == rhs.b and self.a == rhs.a
        )


def CensorActionForFollower(action, follower):
    """Censors actions to hide information that followers aren't supposed to see.

    For now, replaces red border colors with blue.
    """
    if action.border_color == Color(1, 0, 0, 1):
        action = replace(action, border_color=Color(0, 0, 1, 1))
        logger.debug(f"Censored action {action} for follower {follower}")
    return action


@dataclass(frozen=True)
class Action(DataClassJSONMixin):
    id: int = "-1"
    action_type: ActionType = ActionType.NONE
    animation_type: AnimationType = AnimationType.NONE
    # Displacement is used in TRANSLATE, INIT, and INSTANT actions.
    displacement: HecsCoord = HecsCoord.origin()
    # For rotations. In Degrees.
    rotation: float = 0.0
    border_radius: float = 0.0
    border_color: Color = Color(0, 0, 0, 0)
    duration_s: float = 0.0
    expiration: datetime = field(
        metadata={"deserialize": "pendulum", "serialize": pass_through},
        # Default to min time.
        default_factory=lambda: datetime.isoformat(
            datetime.min.replace(tzinfo=tz.tzutc())
        ),
    )
    border_color_follower_pov: Optional[Color] = None  # From follover's point of view.


def Delay(id, duration):
    return Action(
        id=id,
        action_type=ActionType.INSTANT,
        animation_type=AnimationType.INSTANT,
        displacement=HecsCoord.origin(),
        rotation=0,
        border_radius=0,
        border_color=Color(0, 0, 0, 0),
        duration_s=duration,
        expiration=datetime.utcnow() + timedelta(seconds=10),
    )


def Init(id, location, orientation):
    return Action(
        id=id,
        action_type=ActionType.INIT,
        animation_type=AnimationType.IDLE,
        displacement=location,
        rotation=orientation,
        border_radius=0,
        border_color=Color(0, 0, 0, 0),
        duration_s=0.01,
        expiration=datetime.utcnow() + timedelta(seconds=10),
    )


def Turn(id, angle, duration=DEFAULT_TURN_TIME_S):
    return Action(
        id=id,
        action_type=ActionType.ROTATE,
        animation_type=AnimationType.ROTATE,
        displacement=HecsCoord.origin(),
        rotation=angle,
        border_radius=0,
        border_color=Color(0, 0, 0, 0),
        duration_s=duration,
        expiration=datetime.utcnow() + timedelta(seconds=10),
    )


def Walk(id, displacement, duration=DEFAULT_WALK_TIME_S):
    return Action(
        id=id,
        action_type=ActionType.TRANSLATE,
        animation_type=AnimationType.WALKING,
        displacement=displacement,
        rotation=0,
        border_radius=0,
        border_color=Color(0, 0, 0, 0),
        duration_s=duration,
        expiration=datetime.utcnow() + timedelta(seconds=10),
    )
